tasm-lib/src/list/higher_order/map.rs

```diff
@@ -1,330 +1,365 @@
-        let element_type = self.f.range();
-        let list_type = DataType::List(Box::new(element_type));
+        let list_type = DataType::List(Box::new(self.f.range()));
         vec![(list_type, "*output_list".to_string())]
     }
 
     fn entrypoint(&self) -> String {
-        format!("tasmlib_list_higher_order_u32_map_{}", self.f.entrypoint())
+        let maybe_chain_surely_map = if NUM_INPUT_LISTS == 1 {
+            "map".to_string()
+        } else {
+            format!("chain_map_{NUM_INPUT_LISTS}")
+        };
+
+        let f_label = self.f.entrypoint();
+        format!("tasmlib_list_higher_order_u32_{maybe_chain_surely_map}_{f_label}")
     }
 
     fn code(&self, library: &mut Library) -> Vec<LabelledInstruction> {
         let input_type = self.f.domain();
         let output_type = self.f.range();
         let new_list = library.import(Box::new(New::new(output_type.clone())));
 
-        // Declare the code to call the function argument
-        let (call_inner_function, maybe_inner_function_body_raw) = match &self.f {
-            InnerFunction::RawCode(rc) => {
-                if let Some(inlined_body) = rc.inlined_body() {
-                    // Raw functions that can be inlined are inlined to save two clock cycle counts
-                    // for each iteration.
-                    (inlined_body, String::default())
-                } else {
-                    // If function was supplied as raw instructions and cannot be inlined, we need
-                    // to append the inner function to the function body. Otherwise, `library`
-                    // handles the imports.
-                    (
-                        triton_asm!(
-                            call {rc.entrypoint()}
-                        ),
-                        rc.function.iter().map(|x| x.to_string()).join("\n"),
-                    )
-                }
+        let exec_or_call_inner_function = match &self.f {
+            InnerFunction::RawCode(code) => {
+                // Inlining saves two clock cycles per iteration. If the function cannot be
+                // inlined, it needs to be appended to the function body.
+                code.inlined_body()
+                    .unwrap_or(triton_asm!(call {code.entrypoint()}))
             }
             InnerFunction::DeprecatedSnippet(sn) => {
                 assert_eq!(1, sn.input_types().len(), "{INNER_FN_INCORRECT_NUM_INPUTS}");
                 let fn_body = sn.function_code(library);
                 let (_, instructions) = tokenize(&fn_body).unwrap();
                 let labelled_instructions = isa::parser::to_labelled_instructions(&instructions);
-                let snippet_name =
-                    library.explicit_import(&sn.entrypoint_name(), &labelled_instructions);
-                (triton_asm!(call { snippet_name }), String::default())
+                let label = library.explicit_import(&sn.entrypoint_name(), &labelled_instructions);
```

Comment:
For some reason, we are using `explicit_import` here. I'm not 100 % sure that's necessary. Maybe we can just use the `import` method directly on the snippet instead?

Reply:
I tried that, but could not get it to work. The main culprit seems to be the trait object `snippet`, which is of type `&Box<dyn BasicSnippet>`. `Library::import()` requires a `Box<dyn BasicSnippet>`. Dereferencing `snippet` does not work because that would constitute a move, but the method only takes `&self`. Similar problems arise when changing the match statement to match `self.f` instead of `&self.f`. Cloning the `snippet` also does not work, because the `Snippet` trait does not have a trait bound on `Clone`. When trying to introduce such a trait bound, all hell breaks lose.

Long story short: it might be possible, but I decided to give up after a few minutes of trying.